lms/validation/_api.py

```diff
@@ -90,10 +90,15 @@ class File(Schema):
             size = fields.Int(required=True)
 
         type = fields.Str(
-            required=True, validate=marshmallow.validate.OneOf(["url", "file"])
+            required=True,
+            validate=marshmallow.validate.OneOf(["url", "file", "vitalsource"]),
         )
         url = fields.Url(required=False, allow_none=True)
+
         file = fields.Nested(File, required=False, allow_none=True)
+        bookID = fields.Str(required=False, allow_none=True)
+        cfi = fields.Str(required=False, allow_none=True)
+        """VitalSource IDs"""
```

Comment:
I think the `bookID` and `cfi` fields _must_ be present if `type` is `"vitalsource"` and _must not_ be present otherwise. We'd need to add that to the `@validates_schema()` method below I think. Let's make sure we have good unit tests for this as it'd be easy to mess up.

It's getting to the point where I feel we have three separate schemas here (one for URLs, one for files, and one for VitalSource) and we should have a meta-schema that tries to validate the request using each of the sub-schemas in turn. I'm not exactly sure how we'd implement that though.

Reply:
I don't think a few conditionals here are to bad, at least for now where different types don't interact with each other.


>  and must not be present otherwise
Not sure about that one I think if the type says "vitalsource" and we have enough information we should accept it even it also sends something in the URL, same for the other types. I'd `be liberal in what you accept` here.